### PR TITLE
GetPublisherActivityFromUrl now responds by callback

### PR DESCRIPTION
**Pull request by jasonrsadler**
Resolves brave/brave-browser#7104

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

All panel functionality test plans should be followed to test for regression:
Panel publisher information,
Tipping via panel,
etc.

See follow up note.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

cc @NejcZdovc @ryanml @masparrow 

cc @kylehickinson iOS will likely need to be modified

cc @gdregalo @SergeyZhukovsky android

### CODE REVIEW

File: components/brave_rewards/browser/rewards_service_browsertest.cc [hunk 1]
```diff
@@ -2126,7 +2126,7 @@ IN_PROC_BROWSER_TEST_F(BraveRewardsBrowserTest,
   rewards_service_->AddObserver(this);
 
   // Enable Rewards
-  EnableRewardsViaCode();
+  EnableRewards();
```
Comment (jasonrsadler) [inline components/brave_rewards/browser/rewards_service_browsertest.cc]: This test using `EnableRewardsViaCode` would not show publisher info on panel. Reverted to `EnableRewards`

File: vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc [hunk 2]
```diff
@@ -254,7 +249,6 @@ void Publisher::SaveVisitInternal(
                           std::bind(&Publisher::onFetchFavIcon,
```
Comment (jasonrsadler) [inline vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc]: One side effect here is that media publishers update in db but not in time to come out to publisher info so we lose favicon on first visit to media publisher. Subsequent visits show correct favicon for verifieds. Recommend follow up to trace race condition.

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 3]
```diff
@@ -710,7 +710,10 @@ void RewardsServiceImpl::OnUnload(SessionID tab_id) {
   if (!Connected())
     return;
 
-  bat_ledger_->OnUnload(tab_id.id(), GetCurrentTimestamp());
+  bat_ledger_->OnUnload(
+      tab_id.id(),
+      GetCurrentTimestamp(),
+      base::DoNothing());
```
Comment (jasonrsadler) [inline components/brave_rewards/browser/rewards_service_impl.cc]: Some of `On`action methods call into media publishers and consequently required a callback. Recommend some other action other than `base::DoNothing` if appropriate.

**Comment by jasonrsadler** [comment]
This 2nd commit fixes the favicon fetching/saving for media publishers with exceptions for Twitch and Vimeo.

Vimeo seems to have a rendering issue (colored block icons) which is a separate issue.

Twitch can put a users panel favicon for verifieds permanently as the twitch favicon if the panel is opened too early. 

I'm going to attach another PR to this to grab Twitch info directly from their GraphQL servers.

File: vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc (hunk 2 above)
Comment (jasonrsadler) [reply vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc]: Superseded by second commit

**Comment by bsclifton** [comment]
Thanks for the PR, @jasonrsadler - I think it's going to be hard to find bandwidth for reviewing this PR at the moment

- There's a lot of files touched
- There aren't any additional tests written to confirm the refactor works
- The issue it solves is a P5, which isn't prioritized work

For those reasons, I'm going to close the PR. Thanks for putting forth the effort

**Pull request closed without merging**